src/rebase.c

```diff
@@ -1327,6 +1327,33 @@ int git_rebase_finish(
 	return error;
 }
 
+static int git_rebase_get_annotated_name(git_annotated_commit **out, const char *name, git_rebase *rebase) {
+	int error;
+	git_reference *ref;
+
+	if (rebase->type != GIT_REBASE_TYPE_MERGE) {
+		git_error_set(GIT_ERROR_REBASE, "invalid rebase type");
+		return -1;
+	}
+
+	if ((error = git_reference_dwim(&ref, rebase->repo, name)))
+		return error;
+
+	error = git_annotated_commit_from_ref(out, rebase->repo, ref);
+
+	git_reference_free(ref);
+
+	return error;
+}
+
+int git_rebase_annotated_orig_head(git_annotated_commit **out, git_rebase *rebase) {
+	return git_rebase_get_annotated_name(out, rebase->orig_head_name, rebase);
```

Comment:
"Generally", I'd expect that kind of accessor to be of the form `assert(s); return s->member;`, so it has stronger guarantees (ie. you can always get back what you put in). Maybe that could be refactored, but as far as I'm concerned what you're doing here is fine.

Reply:
Well, like I said above, we don't really seem to need the OID at all, so maybe its just best to return the `orig_head` and `onto` names as `const char *` and let the caller resolve/dwim the name himself.

Reply:
There's a bit of a problem with that:  we aren't guaranteed to have the ref names in the future.  Right now we only support two types of rebase (merge-style, and an in-memory format).  The in-memory rebase is something we invented, and since it doesn't write to the filesystem it has no compatibility issues with git, so we can guarantee whatever we want there.

`rebase --merge` emulation has this information, it stores the `onto_name` in a file.  But, other types of rebase _don't_.  You can look at `rebase -i` and see that it has `onto` (an OID) but not the name.  (Because _of course_ it does.)  So when we go to emulate interactive rebase, we'll be missing that information.

So I think we should either expose `onto` and `onto_name` as an oid and a `const char *` (respectively)...  annotated commits are another option, but they're _really_ heavy when all you want is a name and an oid (they actually read the ODB for the commit ID, new up a commit, etc).

Same with `orig_head` for similarity, I think.  We could probably expose the "orig head name" as such to avoid people having to know that it's the "orig head" (an oid) vs the "head name" (a string).

I would propose:

```
git_oid *git_rebase_onto(git_rebase *rebase);
const char *git_rebase_onto_name(git_rebase *rebase);
git_oid *git_rebase_orig_head(git_rebase *rebase);
const char *git_rebase_orig_head_name(git_rebase *rebase);
```

You could also have `onto_id` and `orig_head_id` if you thought that made things more expressive?  I don't care whether the API mirrors the filenames or not.

I would expose _both_ though.  You mention that the caller could resolve them, but there's a pathological case where a user might start a rebase and then edit a ref involved in that rebase in the middle of it.  We should return the literal rebase data.  If a user wanted the `git_oid` for the `orig_head` then if they're expected to get that by resolving the branch name, it could be wrong.  Let's give them the literal data to avoid any surprises or confusion.

Reply:
@ethomson thats how it was in the original commit https://github.com/libgit2/libgit2/commit/44dca762ea3cf47ea92c87780e193d0fff13421a

Reply:
> they actually read the ODB for the commit ID, new up a commit, etc

My point was that this cost is payed by the caller, which passes us those annotated commits. 

From an implementer's perspective, I like annotated commits better than a ref/oid pair, especially since I already looked them up at some point (and then there's the context that an annotated commit carries: things like is that a post-fetch merge will be used to generate a merge-commit message). If I had a separate "rebase-operation" class, it wouldn't be able to give me back the object identity it had: I would have to keep track of the initial values passed in my class before the call to `git_rebase`.

I'm fine with the proposal as it stands, and sorry for the nitpicking.